File: src-tauri/src/python.rs
use std::io::Write;
use std::process::{Command, Stdio};
use std::path::PathBuf;

fn resources_dir() -> PathBuf {
    let exe = std::env::current_exe().unwrap();
    let mut dir = exe.parent().unwrap().to_path_buf();
    // In development: ../python/
    // In production (bundled): ../../../Resources/python/
    if cfg!(debug_assertions) {
        dir.pop(); // remove MacOS/
        dir.pop(); // remove ../../
        dir.push("python");
    } else {
        dir.pop(); // remove MacOS/
        dir.push("Resources");
        dir.push("python");
    }
    dir
}

fn find_python(python_dir: &PathBuf) -> String {
    let candidates = [
        python_dir.join("portable").join("bin").join("python3"),
        python_dir.join("portable").join("bin").join("python3.11"),
    ];
    for p in &candidates {
        if p.exists() {
            return p.to_string_lossy().to_string();
        }
    }
    // fallback to system
    for cmd in &["python3.13", "python3.12", "python3.11", "python3"] {
        if Command::new(cmd)
            .arg("--version")
            .stdout(Stdio::null())
            .stderr(Stdio::null())
            .status()
            .is_ok()
        {
            return cmd.to_string();
        }
    }
    "python3".to_string()
}

pub fn run(params: &str) -> Result<String, String> {
    let python_dir = resources_dir();
    let python = find_python(&python_dir);
    let script = python_dir.join("main.py");

    if !script.exists() {
        return Err(format!(
            "Python script not found at: {}",
            script.display()
        ));
    }

    let mut child = Command::new(&python)
        .arg(&script)
        .stdin(Stdio::piped())
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .spawn()
        .map_err(|e| format!("Failed to start Python: {}", e))?;

    if let Some(mut stdin) = child.stdin.take() {
        stdin
            .write_all(params.as_bytes())
            .map_err(|e| format!("Failed to write stdin: {}", e))?;
    }

    let output = child
        .wait_with_output()
        .map_err(|e| format!("Process error: {}", e))?;

    if output.status.success() {
        let stdout =
            String::from_utf8_lossy(&output.stdout).to_string();
        Ok(stdout)
    } else {
        let stderr =
            String::from_utf8_lossy(&output.stderr).to_string();
        Err(stderr)
    }
}

File: electron/python.ts
import { spawn, execSync } from 'child_process'
import { join, dirname } from 'path'
import { app } from 'electron'
import fs from 'fs'

function getPythonDir(): string {
  if (app.isPackaged) {
    return join(process.resourcesPath, 'python')
  }
  return join(__dirname, '../python')
}

function findPython(pythonDir: string): string {
  const candidates = [
    join(pythonDir, 'portable', 'bin', 'python3'),
    join(pythonDir, 'portable', 'bin', 'python3.11'),
    join(pythonDir, 'portable', 'bin', 'python3.12'),
    join(pythonDir, 'portable', 'bin', 'python3.13'),
  ]

  for (const p of candidates) {
    if (fs.existsSync(p)) {
      return p
    }
  }

  const systemCandidates = ['python3.13', 'python3.12', 'python3.11', 'python3']
  for (const cmd of systemCandidates) {
    try {
      execSync(`${cmd} --version`, { stdio: 'ignore' })
      return cmd
    } catch {
      continue
    }
  }

  return 'python3'
}

export async function runPython(params: string): Promise<string> {
  return new Promise((resolve, reject) => {
    const pythonDir = getPythonDir()
    const python = findPython(pythonDir)
    const script = join(pythonDir, 'main.py')

    if (!fs.existsSync(script)) {
      reject(new Error(`Python script not found at: ${script}`))
      return
    }

    const child = spawn(python, [script], {
      stdio: ['pipe', 'pipe', 'pipe'],
      env: {
        ...process.env,
        PYTHONPATH: pythonDir,
      },
    })

    let stdout = ''
    let stderr = ''

    child.stdout.on('data', (data) => {
      stdout += data.toString()
    })

    child.stderr.on('data', (data) => {
      stderr += data.toString()
    })

    child.stdin.write(params)
    child.stdin.end()

    child.on('close', (code) => {
      if (code === 0) {
        resolve(stdout.trim())
      } else {
        reject(new Error(stderr || `Python process exited with code ${code}`))
      }
    })

    child.on('error', (err) => {
      reject(new Error(`Failed to start Python: ${err.message}`))
    })
  })
}
